### PR TITLE
[SPARK-14480][SQL] Simplify CSV parsing process with a better performance

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-14480

In `CSVParser.scala`, there is an `Reader` wrapping `Iterator` but there are two problems by this.

Firstly, it was actually not faster than processing line by line with Iterator due to additional logics to wrap `Iterator` to `Reader`.
Secondly, this brought a bit of complexity because it needs additional logics to allow every line to be read bytes by bytes. So, it was pretty difficult to figure out issues about parsing, (eg. SPARK-14103).

To cut in short, this PR includes two changes,
- Parse CSV data with `Iterator` not `Reader`.
- Refactor CSV data source to be consistent with JSON data source.

This PR removes classes `CSVParser` and introduces new classes `UnivocityParser`, `UnivocityGenerator` and `CSVUtils` to be consistent with JSON data source (`JacksonParser`, `JacksonGenerator` and `JacksonUtils`). Also, `DefaultSource` moves to `CSVRelation` just like `JSONRelation`.
## How was this patch tested?

Existing tests should cover this.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I have two things to note.
- It looks `buildInternalScan()` is not called but just remaining. Just in case I did not remove this and tested this. 
- It looks the changes are a lot but the logics and filtering are not changed much except that it uses `Iterator` instead of `Reader`.
- Performance was tested with this patch. This can be found in https://issues.apache.org/jira/browse/SPARK-14480.

**Comment by HyukjinKwon** [comment]
cc @rxin

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Could I maybe cc @liancheng  and @cloud-fan to review? This resembles JSON data source structure. So, the class structures and input/output in methods are consistent with JSON data source one.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVParserSuite.scala [hunk 1]
```diff
@@ -1,125 +0,0 @@
...
- */
-
-package org.apache.spark.sql.execution.datasources.csv
-
-import org.apache.spark.SparkFunSuite
-
-/**
- * test cases for StringIteratorReader
- */
-class CSVParserSuite extends SparkFunSuite {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVParserSuite.scala]: Should we add a new parser suite for csv?

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVParserSuite.scala]: Actually, I was a bit worried (and confused) of the range for tests. `CSVParserSuite` was testing the conversion from `Iterator` to `Reader` which does not exist now. So, I was thinking test codes in `UnivocityParser` and `UnivocityGenerator` because they are new classes anyway.

But while trying to add some mode test codes, I realised that it ends up with an end-to-end just as below:
```scala
test("Parse csv data correctly with univocity parser") {
  val cars = sparkContext.textFile(testFile("cars.csv"))
  val schema = StructType(
    StructField("year", IntegerType) ::
    StructField("make", StringType) ::
    StructField("model", StringType) ::
    StructField("comment", StringType) ::
    StructField("blank", StringType) :: Nil)

  val requiredSchema = StructType(
    StructField("year", IntegerType) ::
    StructField("make", StringType) ::
    StructField("model", StringType) :: Nil)

  val headers = schema.fields.map(_.name)
  val options = new CSVOptions(Map.empty)
  val filteredCars = cars.filter(_.nonEmpty)
  val firstLine = filteredCars.first()
  val dropHeaderCars = cars.filter(_ != firstLine)

  val parsedCars = UnivocityParser.parse(
    dropHeaderCars,
    schema,
    requiredSchema,
    headers,
    options).collect()

  val expectedCars = Seq(
    new GenericMutableRow(
      Array(2012, UTF8String.fromString("Tesla"), UTF8String.fromString("S"))),
    new GenericMutableRow(
      Array(1997, UTF8String.fromString("Ford"), UTF8String.fromString("E350"))),
    new GenericMutableRow(
      Array(2015, UTF8String.fromString("Chevy"), UTF8String.fromString("Vold"))))

  assert(parsedCars === expectedCars)
}
```

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVParserSuite.scala]: I just noticed that JSON data source also does not have some tests for `JacksonParser` and `JacksonGenerator`. Could I maybe add some tests for `JacksonParser`, `JacksonGenerator`, `UnivocityParser and UnivocityGenerator in another PR or a follow-up if you think it needs?

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVParserSuite.scala]: yea makes sense, we can do it in follow-ups

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 2]
```diff
@@ -17,153 +17,197 @@
...
-      schema: StructType,
-      requiredColumns: Array[String],
-      params: CSVOptions): Array[String] => Option[InternalRow] = {
-    val schemaFields = schema.fields
-    val requiredFields = StructType(requiredColumns.map(schema(_))).fields
-    val safeRequiredFields = if (params.dropMalformed) {
-      // If `dropMalformed` is enabled, then it needs to parse all the values
-      // so that we can decide which row is malformed.
-      requiredFields ++ schemaFields.filterNot(requiredFields.contains(_))
+    // TODO: Move filtering.
```
Comment (holdenk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Maybe unify this TODO into the same format as the one in JSON relation which is a bit clearer (e.g. "TODO: Filter files for all formats before calling buildInternalScan.")) assuming this is the same filtering you are referring to? Also maybe create a JIRA (or link to existing) since TODOs comments are easy to loose track of.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Thanks! Actually the comment was created by another guy. Anyway I will create some JIRAs for todos I just made.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Actually, let me just leave it as it is. I am not 100% sure if that means what you just said.

Comment (holdenk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: ok - it might be worth while figuring out what this TODO is and clarifying while we are changing the code around it anyways but I can understand just leaving as is.

**Comment by HyukjinKwon** [comment]
@yhuai Could you please review this? I don't want to keep resolving conflicts and I am pretty sure that this is a sensible PR. 

This PR touches pretty a lot of files so this causes many conflicts. I would appreciate that if this is reviewed quicker.

**Comment by cloud-fan** [comment]
cc @rxin , do you know who is the original author of the CSV part?

**Comment by HyukjinKwon** [comment]
Could I please cc @falaki who I believe is the original author?

**Comment by HyukjinKwon** [comment]
Please excuse my pings, @cloud-fan , @rxin , @falaki , @yhuai

**Comment by rxin** [comment]
will try to take a look in the next few days.

**Comment by HyukjinKwon** [comment]
@rxin Could you please review this?

**Comment by HyukjinKwon** [comment]
ping @rxin

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
This was due to https://github.com/apache/spark/commit/d2614eaadb93a48fba27fe7de64aff942e345f8e

**Comment by HyukjinKwon** [comment]
@rxin It looks this is still failing, https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56962
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56963

**Comment by HyukjinKwon** [comment]
Fixed in https://github.com/apache/spark/commit/f8709218115f6c7aa4fb321865cdef8ceb443bd1

**Comment by rxin** [comment]
cc @hvanhovell would you have some time to take a look at this?

@HyukjinKwon most of us are very busy trying to get things out for 2.0 so this one will very likely slip.

**Comment by HyukjinKwon** [comment]
@rxin No problem. Let me just rebase it if it has conflicts anyway. It is easier to track the changes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/InferSchema.scala [hunk 3]
```diff
@@ -30,22 +30,37 @@ import org.apache.spark.sql.catalyst.util.DateTimeUtils
...
    * Similar to the JSON schema inference
    *     1. Infer type of each row
    *     2. Merge row types to find common type
    *     3. Replace any null types with string type
    */
-  def infer(
-      tokenRdd: RDD[Array[String]],
-      header: Array[String],
-      nullValue: String = ""): StructType = {
+  def infer(csv: RDD[String], options: CSVOptions): StructType = {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/InferSchema.scala]: This looks very similar to `DefaultSource.inferSchema` why not move the common functionality  into a single method?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import com.univocity.parsers.csv.{CsvWriter, CsvWriterSettings}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Converts a sequence of string to CSV string
+ */
+private[csv] object UnivocityGenerator extends Logging {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: Are we ever going to use a different generator? Why not call it `CsvGenerator`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+   * Transforms a single InternalRow to CSV using Univocity
+   *
+   * @param rowSchema the schema object used for conversion
+   * @param writer a CsvWriter object
+   * @param headers headers to write
+   * @param writeHeader true if it needs to write header
+   * @param options CSVOptions object containing options
+   * @param row The row to convert
+   */
+  def apply(
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: Please use a more descriptive name? `writeToCsv`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.GenericMutableRow
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Converts CSV string to a sequence of string
+ */
+private[csv] object UnivocityParser extends Logging {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Again naming. At least add csv to the name.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 7]
```diff
@@ -17,152 +17,162 @@
...
-      // so that we can decide which row is malformed.
-      requiredFields ++ schemaFields.filterNot(requiredFields.contains(_))
+  override def inferSchema(
+      sparkSession: SparkSession,
+      options: Map[String, String],
+      files: Seq[FileStatus]): Option[StructType] = {
+    val csvOptions = new CSVOptions(options)
+
+    // TODO: Move filtering.
+    val paths = files.filterNot(_.getPath.getName startsWith "_").map(_.getPath.toString)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: code style: `_.getPath.getName.startsWith("_")`?

Is it safe to skip all files with an underscore?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 8]
```diff
@@ -17,152 +17,162 @@
...
-    tokenizedRDD.flatMap(parser(_).toSeq)
+  override def prepareWrite(
+      sparkSession: SparkSession,
+      job: Job,
+      options: Map[String, String],
+      dataSchema: StructType): OutputWriterFactory = {
+    val conf = job.getConfiguration
+    val csvOptions = new CSVOptions(options)
+    csvOptions.compressionCodec.foreach { codec =>
+      CompressionCodecs.setCodecConfiguration(conf, codec)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Just out of curiosity can we also read compressed csv files?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+      rowSchema: StructType,
+      writer: CsvWriter,
+      headers: Array[String],
+      writeHeader: Boolean,
+      options: CSVOptions)(row: InternalRow): Unit = {
+    val tokens = {
+      row.toSeq(rowSchema).map { field =>
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: You are calling this alot right? So it might be better not to rely on functional constructs here. Also take a look at the `InternalRow.toSeq` method there might be some room improvement if you just pass in the `DataType`s directly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: Come to think of it, why not integrate this with the `CsvOutputWriter`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+            None
+        }
+      }
+    }
+  }
+
+  /**
+   * Convert the tokens to [[InternalRow]]
+   */
+  private def convertTokens(
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: This might be a wild idea: We might be able to use an encoder here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+      tokens: Array[String],
+      requiredIndices: Array[Int],
+      schemaFields: Array[StructField],
+      requiredSize: Int,
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Can an entry in `requiredIndices` lie outside of the `requiredSize` range? Why?

Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Nevermind I got it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    tokenizeData(csv, options, headers).flatMap { tokens =>
+      if (options.dropMalformed && schemaFields.length != tokens.length) {
+        logWarning(s"Dropping malformed line: ${tokens.mkString(options.delimiter.toString)}")
+        None
+      } else if (options.failFast && schemaFields.length != tokens.length) {
+        throw new RuntimeException(s"Malformed line in FAILFAST mode: " +
+          s"${tokens.mkString(options.delimiter.toString)}")
+      } else {
+        val indexSafeTokens = if (options.permissive && schemaFields.length > tokens.length) {
+          tokens ++ new Array[String](schemaFields.length - tokens.length)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Do you think there is way we can do this without appending an array? Using an extra limit in `convertTokens` is probably quicker and causes less GC.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        } else if (options.permissive && schemaFields.length < tokens.length) {
+          tokens.take(schemaFields.length)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Why do we want this? `convertTokens` can't read beyond the `schemaFields.length` right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * Convert the tokens to [[InternalRow]]
+   */
+  private def convertTokens(
+      tokens: Array[String],
+      requiredIndices: Array[Int],
+      schemaFields: Array[StructField],
+      requiredSize: Int,
+      options: CSVOptions): InternalRow = {
+    val row = new GenericMutableRow(requiredSize)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: I am not sure about datasources, but in a lot of places within SparkSQL we just return update a single row and return that over and over.

**Comment by hvanhovell** [comment]
@HyukjinKwon I have taken a pass. The PR looks pretty solid. I do think we can make it a bit more concise in some places and I do think we can make a bit faster as well. Let me know what you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/InferSchema.scala (hunk 3 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/InferSchema.scala]: Actually, it does call this class method in `DefaultSource.inferSchema`. I intentionally made the same structure with `JSONRelation`. JSON data source also have the class with the same name and same method in order to fix issues easily in the future together . (Actually, the main purpose for refactoring this is inconsistency of structures although they could almost identical structures).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala (hunk 4 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: Thanks! The name was also taken after from JSON data source, `JacksonGenerator`. Maybe I can rename them together if this one is merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala (hunk 5 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: The name was also taken after from JSON data source, `JacksonGenerator`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 6 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: The name was also taken after from JSON data source, `JacksonParser`. Can I rename them together with JSON data source if this looks problematic in a follow-up or another PR if this is  sensible?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala (hunk 7 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: I see, I cannot guarantee. JSON data source also skip `name.startsWith("_") || name.startsWith(".")` Let me follow this first. Can I maybe do this together with JSON data source after figuring out in a separate PR or a follow-up?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala (hunk 8 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Oh, yes it is possible since it uses `TextInputFormat`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala (hunk 9 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityGenerator.scala]: Thank you! Could I maybe do this in a separate PR with the purpose of this? This was just copied from original codes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 10 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: I see. Could I do this as well in a separate PR with the purpose of this? Codes were just copied from the original and I just made a function to separate this with the consistent name with JSON data source `convertXXXXX()`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 12 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Thanks for pointing this out. I will think about this further. Maybe I could do this in a separate PR if you think it is sensible. The codes were copied from original.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 13 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Oh, I haven't tested this yet but I am sure this will work without this logic anyway but I think it is safe to slice this here.

The size of `tokens` can be larger than `schemaFields`. I can remove this logic if you feel strongly weird but I feel like it might be okay to just leave.

**Comment by HyukjinKwon** [comment]
@hvanhovell Thank you for a close look! I think I need to change this title of this issue and JIRA because "better performance" might be too broad.

The main purpose of this PR was,
 - Refactoring this to be consistent with JSON data source
 - Remove unnecessary conversion from `Iterator` to `Reader`.

Could I please make some JIRAs and PRs for this in separate PRs or follow-ups if it makes sense?

**Comment by HyukjinKwon** [comment]
@hvanhovell If you think it makes sense I will change the title of this PR and JIRA, and will add some more commits to deal with minor things (code style and etc.).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala (hunk 14 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/UnivocityParser.scala]: Oh yes! I noticed this too. JSON data source will does as far as I remember. This might have to be definitely changed. I think I could do this in another PR.

**Comment by HyukjinKwon** [comment]
@rxin, @hvanhovell Do you mind if I ask your thoughts on this please?

**Comment by rxin** [comment]
Since this is almost a complete rewrite, I think we should only consider it early in the release cycle, i.e. for 2.1, not for 2.0 when we are so close.

**Comment by rxin** [comment]
@HyukjinKwon i just took another look at this. Can you separate this into multiple, smaller pull requests?

IIUC, the main change you want to do is to get rid of the StringIteratorReader. Why don't you just submit a patch for only that? It'd make it much easier to review and merge. The problem with mixing multiple things into a single pull request is that it significantly increases the overhead to review, and makes it almost impossible to merge.

We can start with this one change, and then do other refactorings later.

Also it would be great to show some benchmark numbers since it is a performance improvement. The other thing I'd like to see (in a separate pull request) is to try use the Record API from univocity and compare the performance with our current data cast API.

**Comment by HyukjinKwon** [comment]
@rxin I see. Actually, I added a benchmark for this change in the JIRA.
So.. would this be okay if I do as below:

1. Get rid of `StringIteratorReader`
2. Refactoring (maybe with a follow-up for the tests for.. `JacksonParser`, `JacksonGenerator`, `UnivocityParser` and `UnivocityGenerator`) 
3. Use Record API from univocity and compare the performance (with a proper benchmark)
4. Maybe other fixes addressing @hvanhovell's comments (maybe with a benchmark)

If it makes sense, I would appreciate that If I can create a umbrella for those four as sub-tasts and be assigned to them.

**Comment by rxin** [comment]
Sounds good. Please be surgical in each pr.

Would also be great to include benchmark results in pr description. Thanks.

**Comment by rxin** [comment]
Maybe close this one for now?

**Comment by HyukjinKwon** [comment]
Sure.
